src/memory_region.rs

```diff
@@ -33,52 +33,93 @@ use std::{
 */
 
 /// Callback executed before generate_access_violation()
-pub type MemoryCowCallback = Box<dyn Fn(u32) -> Result<u64, ()>>;
+///
+/// Safety: Reentrancy is forbidden.
+pub type AccessViolationHandler = Box<dyn Fn(&mut MemoryRegion, u64, AccessType, u64, u64)>;
 /// Fail always
 #[allow(clippy::result_unit_err)]
-pub fn default_memory_cow_callback(_cow_callback_payload: u32) -> Result<u64, ()> {
-    Err(())
+pub fn default_access_violation_handler(
+    _region: &mut MemoryRegion,
+    _region_max_len: u64,
+    _access_type: AccessType,
+    _vm_addr: u64,
+    _len: u64,
+) {
+}
+macro_rules! access_violation_guard {
+    ($self_ty:ty, $self:expr, $access_type:expr, $vm_addr:expr, $len:expr) => {{
+        if let Some((index, region)) = $self.find_region($vm_addr) {
+            if let Some(host_addr) = region.vm_to_host($access_type, $vm_addr, $len) {
+                return ProgramResult::Ok(host_addr);
+            }
+            {
+                // Safety: The RefCell prevents reentrancy in AccessViolationHandler.
+                let access_violation_handler = $self.access_violation_handler.borrow_mut();
+                let max_len = $self
+                    .regions
+                    .get(index.saturating_add(1))
+                    .map_or(u64::MAX, |next_region| next_region.vm_addr)
+                    .saturating_sub(region.vm_addr);
+                let mut region = (*region).clone();
+                (&access_violation_handler)(&mut region, max_len, $access_type, $vm_addr, $len);
+                let mut_self = unsafe {
+                    // Same as: &mut *(&raw const *$self).cast_mut()
+                    &mut *(*($self as *const $self_ty).cast::<UnsafeCell<$self_ty>>()).get()
+                };
```

Comment:
For the sake of safe code, would it be too hard to make `fn map` to receive `&mut self`?

Reply:
Yes, thought about it too. It requires quite a few changes in the Agave repo, but I will see what I can do.